src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/PoolingDataSource.cs

```diff
@@ -424,13 +428,46 @@ internal void PruneIdleConnections()
             throw new NotImplementedException();
         }
 
+        private void QueueWarmupTask()
+        {
+            // Make sure we're at quota by posting a callback to the threadpool.
+            ThreadPool.QueueUserWorkItem(_warmupCallback);
+        }
+
         /// <summary>
         /// Warms up the pool to bring it up to min pool size.
         /// </summary>
         /// <exception cref="NotImplementedException"></exception>
-        internal void WarmUp()
+        private async void WarmupCallback(object? state)
         {
-            throw new NotImplementedException();
+            /* Best effort, we may create at most one unneeded connection.
+             * 
+             * Open new connections slowly. If many connections are needed immediately 
+             * upon pool creation they can always be created via user-initiated requests as fast
+             * as a parallel, pool-initiated approach could.
+             */
+            while (_numConnectors < MinPoolSize)
+            {
+                // Obey the same rate limit as user-initiated opens.
+                // Ensures that pool-initiated opens are queued properly alongside user requests.
+                SqlConnector? connector = await OpenNewInternalConnection(
+                    null,
+                    TimeSpan.FromSeconds(Settings.ConnectTimeout),
+                    true,
+                    CancellationToken.None
+                ).ConfigureAwait(false);
```

Comment:
I personally like this style of putting the closing paren on a new line, but MSFT C# style seems to be to leave it on the same line as the last parameter. The `.ConfigureAwait` would go on the next line, I think...

```
var connector = await OpenNewInternalConnection(
    ...
    CancellationToken.None)
    .ConfigureAwait(false);
```